<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>historydag.mutation_annotated_dag &mdash; historydag  documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js?v=5d32c60e"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../_static/documentation_options.js?v=5929fcd5"></script>
        <script src="../_static/doctools.js?v=888ff710"></script>
        <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="historydag.utils" href="historydag.utils.html" />
    <link rel="prev" title="historydag.sequence_dag" href="historydag.sequence_dag.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html">
            
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">User Guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../quickstart.html">Quickstart</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API Documentation</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../api.html">Description</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#classes">Classes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#functions">Functions</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../api.html#modules">Modules</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="historydag.dag.html">historydag.dag</a></li>
<li class="toctree-l2"><a class="reference internal" href="historydag.sequence_dag.html">historydag.sequence_dag</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">historydag.mutation_annotated_dag</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#historydag.mutation_annotated_dag.HDagJSONEncoder"><code class="docutils literal notranslate"><span class="pre">HDagJSONEncoder</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#historydag.mutation_annotated_dag.HDagJSONEncoder.default"><code class="docutils literal notranslate"><span class="pre">HDagJSONEncoder.default()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#historydag.mutation_annotated_dag.NodeIDHistoryDag"><code class="docutils literal notranslate"><span class="pre">NodeIDHistoryDag</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#historydag.mutation_annotated_dag.CGHistoryDag"><code class="docutils literal notranslate"><span class="pre">CGHistoryDag</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#historydag.mutation_annotated_dag.CGHistoryDag.weight_counts_with_ambiguities"><code class="docutils literal notranslate"><span class="pre">CGHistoryDag.weight_counts_with_ambiguities()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#historydag.mutation_annotated_dag.CGHistoryDag.summary"><code class="docutils literal notranslate"><span class="pre">CGHistoryDag.summary()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#historydag.mutation_annotated_dag.CGHistoryDag.hamming_parsimony_count"><code class="docutils literal notranslate"><span class="pre">CGHistoryDag.hamming_parsimony_count()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#historydag.mutation_annotated_dag.CGHistoryDag.to_protobuf"><code class="docutils literal notranslate"><span class="pre">CGHistoryDag.to_protobuf()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#historydag.mutation_annotated_dag.CGHistoryDag.to_protobuf_file"><code class="docutils literal notranslate"><span class="pre">CGHistoryDag.to_protobuf_file()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#historydag.mutation_annotated_dag.CGHistoryDag.flatten"><code class="docutils literal notranslate"><span class="pre">CGHistoryDag.flatten()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#historydag.mutation_annotated_dag.CGHistoryDag.test_equal"><code class="docutils literal notranslate"><span class="pre">CGHistoryDag.test_equal()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#historydag.mutation_annotated_dag.CGHistoryDag.get_reference_sequence"><code class="docutils literal notranslate"><span class="pre">CGHistoryDag.get_reference_sequence()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#historydag.mutation_annotated_dag.CGHistoryDag.to_json"><code class="docutils literal notranslate"><span class="pre">CGHistoryDag.to_json()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#historydag.mutation_annotated_dag.CGHistoryDag.to_json_file"><code class="docutils literal notranslate"><span class="pre">CGHistoryDag.to_json_file()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#historydag.mutation_annotated_dag.CGHistoryDag.adjusted_node_probabilities"><code class="docutils literal notranslate"><span class="pre">CGHistoryDag.adjusted_node_probabilities()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#historydag.mutation_annotated_dag.AmbiguousLeafCGHistoryDag"><code class="docutils literal notranslate"><span class="pre">AmbiguousLeafCGHistoryDag</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#historydag.mutation_annotated_dag.AmbiguousLeafCGHistoryDag.hamming_parsimony_count"><code class="docutils literal notranslate"><span class="pre">AmbiguousLeafCGHistoryDag.hamming_parsimony_count()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#historydag.mutation_annotated_dag.AmbiguousLeafCGHistoryDag.summary"><code class="docutils literal notranslate"><span class="pre">AmbiguousLeafCGHistoryDag.summary()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#historydag.mutation_annotated_dag.load_json_file"><code class="docutils literal notranslate"><span class="pre">load_json_file()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#historydag.mutation_annotated_dag.unflatten"><code class="docutils literal notranslate"><span class="pre">unflatten()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#historydag.mutation_annotated_dag.load_MAD_protobuf"><code class="docutils literal notranslate"><span class="pre">load_MAD_protobuf()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#historydag.mutation_annotated_dag.load_MAD_protobuf_file"><code class="docutils literal notranslate"><span class="pre">load_MAD_protobuf_file()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="historydag.utils.html">historydag.utils</a></li>
<li class="toctree-l2"><a class="reference internal" href="historydag.parsimony.html">historydag.parsimony</a></li>
<li class="toctree-l2"><a class="reference internal" href="historydag.compact_genome.html">historydag.compact_genome</a></li>
<li class="toctree-l2"><a class="reference internal" href="historydag.counterops.html">historydag.counterops</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">historydag</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../api.html">Description</a></li>
      <li class="breadcrumb-item active">historydag.mutation_annotated_dag</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-historydag.mutation_annotated_dag">
<span id="historydag-mutation-annotated-dag"></span><h1>historydag.mutation_annotated_dag<a class="headerlink" href="#module-historydag.mutation_annotated_dag" title="Link to this heading"></a></h1>
<p>This module allows the loading and manipulation of Larch mutation annotated
DAG protobuf files.</p>
<p>The resulting history DAG contains labels with ‘compact genomes’, and a
‘refseq’ attribute describing a reference sequence and set of mutations
relative to the reference.</p>
<p class="rubric">Functions</p>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#historydag.mutation_annotated_dag.load_MAD_protobuf" title="historydag.mutation_annotated_dag.load_MAD_protobuf"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_MAD_protobuf</span></code></a>(pbdata[, compact_genomes, ...])</p></td>
<td><p>Convert a Larch MAD protobuf to a CGLeafIDHistoryDag with compact genomes in the <cite>compact_genome</cite> label attribute.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#historydag.mutation_annotated_dag.load_MAD_protobuf_file" title="historydag.mutation_annotated_dag.load_MAD_protobuf_file"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_MAD_protobuf_file</span></code></a>(filename, **kwargs)</p></td>
<td><p>Load a mutation annotated DAG protobuf file and return a CGHistoryDag.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#historydag.mutation_annotated_dag.load_json_file" title="historydag.mutation_annotated_dag.load_json_file"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_json_file</span></code></a>(filename)</p></td>
<td><p>Load a Mutation Annotated DAG stored in a JSON file and return a CGHistoryDag.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#historydag.mutation_annotated_dag.unflatten" title="historydag.mutation_annotated_dag.unflatten"><code class="xref py py-obj docutils literal notranslate"><span class="pre">unflatten</span></code></a>(flat_dag)</p></td>
<td><p>Takes a dictionary like that returned by flatten, and returns a HistoryDag.</p></td>
</tr>
</tbody>
</table>
<p class="rubric">Classes</p>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#historydag.mutation_annotated_dag.AmbiguousLeafCGHistoryDag" title="historydag.mutation_annotated_dag.AmbiguousLeafCGHistoryDag"><code class="xref py py-obj docutils literal notranslate"><span class="pre">AmbiguousLeafCGHistoryDag</span></code></a>(dagroot[, attr])</p></td>
<td><p>A HistoryDag subclass with node labels containing compact genomes.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#historydag.mutation_annotated_dag.CGHistoryDag" title="historydag.mutation_annotated_dag.CGHistoryDag"><code class="xref py py-obj docutils literal notranslate"><span class="pre">CGHistoryDag</span></code></a>(dagroot[, attr])</p></td>
<td><p>A HistoryDag subclass with node labels containing CompactGenome objects.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#historydag.mutation_annotated_dag.HDagJSONEncoder" title="historydag.mutation_annotated_dag.HDagJSONEncoder"><code class="xref py py-obj docutils literal notranslate"><span class="pre">HDagJSONEncoder</span></code></a>(*[, skipkeys, ensure_ascii, ...])</p></td>
<td><p>Constructor for JSONEncoder, with sensible defaults.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#historydag.mutation_annotated_dag.NodeIDHistoryDag" title="historydag.mutation_annotated_dag.NodeIDHistoryDag"><code class="xref py py-obj docutils literal notranslate"><span class="pre">NodeIDHistoryDag</span></code></a>(dagroot[, attr])</p></td>
<td><p>A HistoryDag subclass with node labels containing string <code class="docutils literal notranslate"><span class="pre">node_id</span></code> fields.</p></td>
</tr>
</tbody>
</table>
<dl class="py class">
<dt class="sig sig-object py" id="historydag.mutation_annotated_dag.HDagJSONEncoder">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">historydag.mutation_annotated_dag.</span></span><span class="sig-name descname"><span class="pre">HDagJSONEncoder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">skipkeys</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensure_ascii</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check_circular</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sort_keys</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">indent</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">separators</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">default</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/historydag/mutation_annotated_dag.html#HDagJSONEncoder"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#historydag.mutation_annotated_dag.HDagJSONEncoder" title="Link to this definition"></a></dt>
<dd><p>Constructor for JSONEncoder, with sensible defaults.</p>
<p>If skipkeys is false, then it is a TypeError to attempt
encoding of keys that are not str, int, float or None.  If
skipkeys is True, such items are simply skipped.</p>
<p>If ensure_ascii is true, the output is guaranteed to be str
objects with all incoming non-ASCII characters escaped.  If
ensure_ascii is false, the output can contain non-ASCII characters.</p>
<p>If check_circular is true, then lists, dicts, and custom encoded
objects will be checked for circular references during encoding to
prevent an infinite recursion (which would cause an RecursionError).
Otherwise, no such check takes place.</p>
<p>If allow_nan is true, then NaN, Infinity, and -Infinity will be
encoded as such.  This behavior is not JSON specification compliant,
but is consistent with most JavaScript based encoders and decoders.
Otherwise, it will be a ValueError to encode such floats.</p>
<p>If sort_keys is true, then the output of dictionaries will be
sorted by key; this is useful for regression tests to ensure
that JSON serializations can be compared on a day-to-day basis.</p>
<p>If indent is a non-negative integer, then JSON array
elements and object members will be pretty-printed with that
indent level.  An indent level of 0 will only insert newlines.
None is the most compact representation.</p>
<p>If specified, separators should be an (item_separator, key_separator)
tuple.  The default is (’, ‘, ‘: ‘) if <em>indent</em> is <code class="docutils literal notranslate"><span class="pre">None</span></code> and
(‘,’, ‘: ‘) otherwise.  To get the most compact JSON representation,
you should specify (‘,’, ‘:’) to eliminate whitespace.</p>
<p>If specified, default is a function that gets called for objects
that can’t otherwise be serialized.  It should return a JSON encodable
version of the object or raise a <code class="docutils literal notranslate"><span class="pre">TypeError</span></code>.</p>
<dl class="py method">
<dt class="sig sig-object py" id="historydag.mutation_annotated_dag.HDagJSONEncoder.default">
<span class="sig-name descname"><span class="pre">default</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">obj</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/historydag/mutation_annotated_dag.html#HDagJSONEncoder.default"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#historydag.mutation_annotated_dag.HDagJSONEncoder.default" title="Link to this definition"></a></dt>
<dd><p>Implement this method in a subclass such that it returns
a serializable object for <code class="docutils literal notranslate"><span class="pre">o</span></code>, or calls the base implementation
(to raise a <code class="docutils literal notranslate"><span class="pre">TypeError</span></code>).</p>
<p>For example, to support arbitrary iterators, you could
implement default like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">default</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">o</span><span class="p">):</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">iterable</span> <span class="o">=</span> <span class="nb">iter</span><span class="p">(</span><span class="n">o</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">TypeError</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="n">iterable</span><span class="p">)</span>
    <span class="c1"># Let the base class default method raise the TypeError</span>
    <span class="k">return</span> <span class="n">JSONEncoder</span><span class="o">.</span><span class="n">default</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">o</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="historydag.mutation_annotated_dag.NodeIDHistoryDag">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">historydag.mutation_annotated_dag.</span></span><span class="sig-name descname"><span class="pre">NodeIDHistoryDag</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dagroot</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">attr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/historydag/mutation_annotated_dag.html#NodeIDHistoryDag"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#historydag.mutation_annotated_dag.NodeIDHistoryDag" title="Link to this definition"></a></dt>
<dd><p>A HistoryDag subclass with node labels containing string <code class="docutils literal notranslate"><span class="pre">node_id</span></code>
fields.</p>
<p>For leaf nodes this string is a unique leaf identifier, and for
internal nodes this is a string representation of an integer node
ID.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="historydag.mutation_annotated_dag.CGHistoryDag">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">historydag.mutation_annotated_dag.</span></span><span class="sig-name descname"><span class="pre">CGHistoryDag</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dagroot</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">attr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/historydag/mutation_annotated_dag.html#CGHistoryDag"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#historydag.mutation_annotated_dag.CGHistoryDag" title="Link to this definition"></a></dt>
<dd><p>A HistoryDag subclass with node labels containing CompactGenome objects.</p>
<p>The constructor for this class requires that each node label contain
a ‘compact_genome’ field, which is expected to hold a
<code class="xref py py-class docutils literal notranslate"><span class="pre">compact_genome.CompactGenome</span></code> object.</p>
<p>A HistoryDag containing ‘sequence’ node label fields may be
automatically converted to this subclass by calling the class method
<code class="xref py py-meth docutils literal notranslate"><span class="pre">CGHistoryDag.from_dag()</span></code>, providing the HistoryDag object to be
converted, and the reference sequence to the keyword argument
‘reference’.</p>
<p>This subclass provides specialized methods for interfacing with
Larch’s MADAG protobuf format</p>
<dl class="py method">
<dt class="sig sig-object py" id="historydag.mutation_annotated_dag.CGHistoryDag.weight_counts_with_ambiguities">
<span class="sig-name descname"><span class="pre">weight_counts_with_ambiguities</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/historydag/mutation_annotated_dag.html#CGHistoryDag.weight_counts_with_ambiguities"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#historydag.mutation_annotated_dag.CGHistoryDag.weight_counts_with_ambiguities" title="Link to this definition"></a></dt>
<dd><p>Template method for counting tree weights in the DAG, with exploded
labels. Like <code class="xref py py-meth docutils literal notranslate"><span class="pre">HistoryDag.weight_count()</span></code>, but creates dictionaries
of Counter objects at each node, keyed by possible sequences at that
node. Analogous to <code class="xref py py-meth docutils literal notranslate"><span class="pre">HistoryDag.count_histories()</span></code> with
<cite>expand_func</cite> provided.</p>
<p>Weights must be hashable.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start_func</strong> – A function which assigns a weight to each leaf node</p></li>
<li><p><strong>edge_func</strong> – A function which assigns a weight to pairs of labels, with the
parent node label the first argument. Must correctly handle the UA
node label which is a UALabel instead of a namedtuple.</p></li>
<li><p><strong>accum_func</strong> – A way to ‘add’ a list of weights together</p></li>
<li><p><strong>expand_func</strong> – A function which takes a label and returns a list of labels, such
as disambiguations of an ambiguous sequence.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A Counter keyed by weights.
The total number of trees will be greater than count_histories(), as these are
possible disambiguations of trees. These disambiguations may not be unique,
but if two are the same, they come from different subtrees of the DAG.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="historydag.mutation_annotated_dag.CGHistoryDag.summary">
<span class="sig-name descname"><span class="pre">summary</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/historydag/mutation_annotated_dag.html#CGHistoryDag.summary"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#historydag.mutation_annotated_dag.CGHistoryDag.summary" title="Link to this definition"></a></dt>
<dd><p>Print summary info about the history DAG.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="historydag.mutation_annotated_dag.CGHistoryDag.hamming_parsimony_count">
<span class="sig-name descname"><span class="pre">hamming_parsimony_count</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/historydag/mutation_annotated_dag.html#CGHistoryDag.hamming_parsimony_count"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#historydag.mutation_annotated_dag.CGHistoryDag.hamming_parsimony_count" title="Link to this definition"></a></dt>
<dd><p>See <code class="xref py py-meth docutils literal notranslate"><span class="pre">historydag.sequence_dag.SequenceHistoryDag.hamming_parsim</span>
<span class="pre">ony_count()</span></code></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="historydag.mutation_annotated_dag.CGHistoryDag.to_protobuf">
<span class="sig-name descname"><span class="pre">to_protobuf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">leaf_data_func</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">randomize_leaf_muts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transition_model</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">default_nt_transitions</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/historydag/mutation_annotated_dag.html#CGHistoryDag.to_protobuf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#historydag.mutation_annotated_dag.CGHistoryDag.to_protobuf" title="Link to this definition"></a></dt>
<dd><p>Convert a DAG with compact genome data on each node, and unique leaf
IDs on leaf nodes, to a MAD protobuf with mutation information on
edges.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>leaf_data_func</strong> – a function taking a DAG node and returning a string to store
in the protobuf node_name field <cite>condensed_leaves</cite> of leaf nodes. On leaf
nodes, this data is appended after the unique leaf ID.</p></li>
<li><p><strong>randomize_leaf_muts</strong> – When leaf node sequences contain ambiguities, if
True the mutations on pendant edges will be randomized, when
there are multiple choices.</p></li>
<li><p><strong>transition_model</strong> – A <code class="xref py py-meth docutils literal notranslate"><span class="pre">historydag.parsimony_utils.TransitionModel()</span></code> object,
used to decide which bases to record on pendant edge mutations with
ambiguous bases as targets.</p></li>
</ul>
</dd>
</dl>
<p>Note that internal node IDs will be reassigned, even if internal nodes have
node IDs in their label data.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="historydag.mutation_annotated_dag.CGHistoryDag.to_protobuf_file">
<span class="sig-name descname"><span class="pre">to_protobuf_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">leaf_data_func</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">randomize_leaf_muts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/historydag/mutation_annotated_dag.html#CGHistoryDag.to_protobuf_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#historydag.mutation_annotated_dag.CGHistoryDag.to_protobuf_file" title="Link to this definition"></a></dt>
<dd><p>Write this CGHistoryDag to a Mutation Annotated DAG protobuf for use
with Larch.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="historydag.mutation_annotated_dag.CGHistoryDag.flatten">
<span class="sig-name descname"><span class="pre">flatten</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sort_compact_genomes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/historydag/mutation_annotated_dag.html#CGHistoryDag.flatten"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#historydag.mutation_annotated_dag.CGHistoryDag.flatten" title="Link to this definition"></a></dt>
<dd><p>Return a dictionary containing four keys:</p>
<ul class="simple">
<li><p><cite>refseq</cite> is a list containing the reference sequence id, and the reference sequence
(the implied sequence on the UA node)</p></li>
<li><p><cite>compact_genome_list</cite> is a list of compact genomes, where each compact genome is a
list of nested lists <cite>[seq_idx, [old_base, new_base]]</cite> where <cite>seq_idx</cite> is (1-indexed)
nucleotide sequence site. If sort_compact_genomes is True, compact genomes and <cite>compact_genome_list</cite> are sorted.</p></li>
<li><dl class="simple">
<dt><cite>node_list</cite> is a list of <cite>[label_idx, clade_list]</cite> pairs, where</dt><dd><ul>
<li><p><cite>label_idx</cite> is the index of the node’s compact genome in <cite>compact_genome_list</cite>, and</p></li>
<li><p><cite>clade_list</cite> is a list of lists of <cite>compact_genome_list</cite> indices, encoding sets of child clades.</p></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>edge_list</cite> is a list of triples <cite>[parent_idx, child_idx, clade_idx]</cite>, where</dt><dd><ul>
<li><p><cite>parent_idx</cite> is the index of the edge’s parent node in <cite>node_list</cite>,</p></li>
<li><p><cite>child_idx</cite> is the index of the edge’s child node in <cite>node_list</cite>, and</p></li>
<li><p><cite>clade_idx</cite> is the index of the clade in the parent node’s <cite>clade_list</cite> from which this edge descends.</p></li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="historydag.mutation_annotated_dag.CGHistoryDag.test_equal">
<span class="sig-name descname"><span class="pre">test_equal</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/historydag/mutation_annotated_dag.html#CGHistoryDag.test_equal"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#historydag.mutation_annotated_dag.CGHistoryDag.test_equal" title="Link to this definition"></a></dt>
<dd><p>Deprecated test for whether two history DAGs are equal.</p>
<p>Compares sorted JSON representation. Only works when
“compact_genome” is the only label field, on all nodes.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="historydag.mutation_annotated_dag.CGHistoryDag.get_reference_sequence">
<span class="sig-name descname"><span class="pre">get_reference_sequence</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/historydag/mutation_annotated_dag.html#CGHistoryDag.get_reference_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#historydag.mutation_annotated_dag.CGHistoryDag.get_reference_sequence" title="Link to this definition"></a></dt>
<dd><p>Return the reference sequence for this CGHistoryDag.</p>
<p>This is the sequence with respect to which all node label
CompactGenomes record mutations.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="historydag.mutation_annotated_dag.CGHistoryDag.to_json">
<span class="sig-name descname"><span class="pre">to_json</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sort_compact_genomes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/historydag/mutation_annotated_dag.html#CGHistoryDag.to_json"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#historydag.mutation_annotated_dag.CGHistoryDag.to_json" title="Link to this definition"></a></dt>
<dd><p>Write this history DAG to a JSON object.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="historydag.mutation_annotated_dag.CGHistoryDag.to_json_file">
<span class="sig-name descname"><span class="pre">to_json_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sort_compact_genomes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/historydag/mutation_annotated_dag.html#CGHistoryDag.to_json_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#historydag.mutation_annotated_dag.CGHistoryDag.to_json_file" title="Link to this definition"></a></dt>
<dd><p>Write this history DAG to a JSON file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="historydag.mutation_annotated_dag.CGHistoryDag.adjusted_node_probabilities">
<span class="sig-name descname"><span class="pre">adjusted_node_probabilities</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">log_probabilities</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ua_node_val</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adjust_func</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/historydag/mutation_annotated_dag.html#CGHistoryDag.adjusted_node_probabilities"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#historydag.mutation_annotated_dag.CGHistoryDag.adjusted_node_probabilities" title="Link to this definition"></a></dt>
<dd><p>Compute the probability of each node in the DAG, adjusted based on
the frequency of mutations that define each node.</p>
<p>See <code class="xref py py-meth docutils literal notranslate"><span class="pre">HistoryDag.node_probabilities()</span></code> for argument
descriptions.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="historydag.mutation_annotated_dag.AmbiguousLeafCGHistoryDag">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">historydag.mutation_annotated_dag.</span></span><span class="sig-name descname"><span class="pre">AmbiguousLeafCGHistoryDag</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dagroot</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">attr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/historydag/mutation_annotated_dag.html#AmbiguousLeafCGHistoryDag"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#historydag.mutation_annotated_dag.AmbiguousLeafCGHistoryDag" title="Link to this definition"></a></dt>
<dd><p>A HistoryDag subclass with node labels containing compact genomes.</p>
<p>The constructor for this class requires that each node label contain
a ‘compact_genome’ field, which is expected to hold a
<code class="xref py py-class docutils literal notranslate"><span class="pre">compact_genome.CompactGenome</span></code> object, which is expected to
hold an unambiguous sequence if the node is internal. The sequence
may contain ambiguities if the node is a leaf.</p>
<p>A HistoryDag containing ‘sequence’ node label fields may be
automatically converted to this subclass by calling the class method
<code class="xref py py-meth docutils literal notranslate"><span class="pre">CGHistoryDag.from_dag()</span></code>, providing the HistoryDag object to be
converted, and the reference sequence to the keyword argument
‘reference’.</p>
<dl class="py method">
<dt class="sig sig-object py" id="historydag.mutation_annotated_dag.AmbiguousLeafCGHistoryDag.hamming_parsimony_count">
<span class="sig-name descname"><span class="pre">hamming_parsimony_count</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/historydag/mutation_annotated_dag.html#AmbiguousLeafCGHistoryDag.hamming_parsimony_count"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#historydag.mutation_annotated_dag.AmbiguousLeafCGHistoryDag.hamming_parsimony_count" title="Link to this definition"></a></dt>
<dd><p>See <code class="xref py py-meth docutils literal notranslate"><span class="pre">historydag.sequence_dag.SequenceHistoryDag.hamming_parsim</span>
<span class="pre">ony_count()</span></code></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="historydag.mutation_annotated_dag.AmbiguousLeafCGHistoryDag.summary">
<span class="sig-name descname"><span class="pre">summary</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/historydag/mutation_annotated_dag.html#AmbiguousLeafCGHistoryDag.summary"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#historydag.mutation_annotated_dag.AmbiguousLeafCGHistoryDag.summary" title="Link to this definition"></a></dt>
<dd><p>Print summary info about the history DAG.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="historydag.mutation_annotated_dag.load_json_file">
<span class="sig-prename descclassname"><span class="pre">historydag.mutation_annotated_dag.</span></span><span class="sig-name descname"><span class="pre">load_json_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/historydag/mutation_annotated_dag.html#load_json_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#historydag.mutation_annotated_dag.load_json_file" title="Link to this definition"></a></dt>
<dd><p>Load a Mutation Annotated DAG stored in a JSON file and return a
CGHistoryDag.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="historydag.mutation_annotated_dag.unflatten">
<span class="sig-prename descclassname"><span class="pre">historydag.mutation_annotated_dag.</span></span><span class="sig-name descname"><span class="pre">unflatten</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">flat_dag</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/historydag/mutation_annotated_dag.html#unflatten"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#historydag.mutation_annotated_dag.unflatten" title="Link to this definition"></a></dt>
<dd><p>Takes a dictionary like that returned by flatten, and returns a
HistoryDag.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="historydag.mutation_annotated_dag.load_MAD_protobuf">
<span class="sig-prename descclassname"><span class="pre">historydag.mutation_annotated_dag.</span></span><span class="sig-name descname"><span class="pre">load_MAD_protobuf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pbdata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compact_genomes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">leaf_cgs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ambiguity_map</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">standard_nt_ambiguity_map</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/historydag/mutation_annotated_dag.html#load_MAD_protobuf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#historydag.mutation_annotated_dag.load_MAD_protobuf" title="Link to this definition"></a></dt>
<dd><p>Convert a Larch MAD protobuf to a CGLeafIDHistoryDag with compact
genomes in the <cite>compact_genome</cite> label attribute.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pbdata</strong> – loaded protobuf data object</p></li>
<li><p><strong>compact_genomes</strong> – If True, returns a CGHistoryDag or AmbiguousLeafCGHistoryDag
object, with labels containing <cite>node_id</cite> and <cite>compact_genome</cite> fields.
If no leaf sequence data is provided, leaf compact genomes will be
inferred from pendant edge mutations, and will include ambiguities when
mutations on two pendant edges pointing to the same leaf would otherwise
contradict. <cite>node_id</cite> field on internal nodes will be None, unless
<cite>node_ids</cite> argument is True. If False, this function will return a NodeIDHistoryDag.</p></li>
<li><p><strong>node_ids</strong> – If True, node IDs will be included on all nodes’ labels. If False, internal
nodes’ <cite>node_id</cite> label fields will be <cite>None</cite>. Unique leaf sequence IDs are always
included in the <cite>node_id</cite> label field of leaf nodes, to ensure that leaf node
labels are unique.</p></li>
<li><p><strong>leaf_cgs</strong> – (not implemented) A dictionary keyed by unique string leaf IDs containing CompactGenomes.
Use <code class="xref py py-meth docutils literal notranslate"><span class="pre">compact_genome.read_alignment()</span></code> to read an alignment from a file.</p></li>
<li><p><strong>ambiguity_map</strong> – A <code class="xref py py-meth docutils literal notranslate"><span class="pre">historydag.parsimony_utils.AmbiguityMap()</span></code> object
to determine how conflicting pendant edge mutations are represented.</p></li>
</ul>
</dd>
</dl>
<p>Note that if leaf sequences in the original alignment do not contain ambiguities, it is not
necessary to provide alignment data; leaf sequences can be completely inferred without it.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="historydag.mutation_annotated_dag.load_MAD_protobuf_file">
<span class="sig-prename descclassname"><span class="pre">historydag.mutation_annotated_dag.</span></span><span class="sig-name descname"><span class="pre">load_MAD_protobuf_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/historydag/mutation_annotated_dag.html#load_MAD_protobuf_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#historydag.mutation_annotated_dag.load_MAD_protobuf_file" title="Link to this definition"></a></dt>
<dd><p>Load a mutation annotated DAG protobuf file and return a
CGHistoryDag.</p>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="historydag.sequence_dag.html" class="btn btn-neutral float-left" title="historydag.sequence_dag" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="historydag.utils.html" class="btn btn-neutral float-right" title="historydag.utils" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2020, Will Dumm.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>