<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>historydag.compact_genome &mdash; historydag  documentation</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../../_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../../_static/documentation_options.js?v=5929fcd5"></script>
        <script src="../../_static/doctools.js?v=888ff710"></script>
        <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html">
            
              <img src="../../_static/logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">User Guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../quickstart.html">Quickstart</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../api.html">Description</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api.html#classes">Classes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api.html#functions">Functions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api.html#modules">Modules</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">historydag</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">Module code</a></li>
      <li class="breadcrumb-item active">historydag.compact_genome</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for historydag.compact_genome</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;This module provides a CompactGenome class, intended as a convenient and compact representation of</span>
<span class="sd">a nucleotide sequence as a collection of mutations relative to a reference sequence. This object also</span>
<span class="sd">provides methods to conveniently mutate CompactGenome objects according to a list of mutations, produce</span>
<span class="sd">mutations defining the difference between two CompactGenome objects, and efficiently access the base</span>
<span class="sd">at a site (or the entire sequence, as a string) implied by a CompactGenome.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">from</span> <span class="nn">frozendict</span> <span class="kn">import</span> <span class="n">frozendict</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">Sequence</span>
<span class="kn">from</span> <span class="nn">warnings</span> <span class="kn">import</span> <span class="n">warn</span>
<span class="kn">from</span> <span class="nn">historydag.parsimony_utils</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">standard_nt_ambiguity_map</span><span class="p">,</span>
    <span class="n">default_nt_transitions</span><span class="p">,</span>
    <span class="n">CharacterSequence</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">historydag.utils</span> <span class="kn">import</span> <span class="n">read_fasta</span>


<div class="viewcode-block" id="CompactGenome">
<a class="viewcode-back" href="../../stubs/historydag.compact_genome.html#historydag.compact_genome.CompactGenome">[docs]</a>
<span class="k">class</span> <span class="nc">CompactGenome</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;A collection of mutations relative to a reference sequence.</span>

<span class="sd">    Args:</span>
<span class="sd">        mutations: The difference between the reference and this sequence, expressed</span>
<span class="sd">            in a dictionary, in which keys are one-based sequence indices, and values</span>
<span class="sd">            are (reference base, new base) pairs.</span>
<span class="sd">        reference: The reference sequence</span>
<span class="sd">    &quot;&quot;&quot;</span>

<div class="viewcode-block" id="CompactGenome.__init__">
<a class="viewcode-back" href="../../stubs/historydag.compact_genome.html#historydag.compact_genome.CompactGenome.__init__">[docs]</a>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">mutations</span><span class="p">:</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">reference</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reference</span> <span class="o">=</span> <span class="n">reference</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mutations</span> <span class="o">=</span> <span class="n">frozendict</span><span class="p">(</span><span class="n">mutations</span><span class="p">)</span></div>


    <span class="k">def</span> <span class="fm">__hash__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="nb">hash</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="p">)</span>

    <span class="k">def</span> <span class="fm">__eq__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">):</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="n">CompactGenome</span><span class="p">):</span>
            <span class="k">return</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference</span><span class="p">)</span> <span class="o">==</span> <span class="p">(</span>
                <span class="n">other</span><span class="o">.</span><span class="n">mutations</span><span class="p">,</span>
                <span class="n">other</span><span class="o">.</span><span class="n">reference</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">NotImplementedError</span>

    <span class="k">def</span> <span class="fm">__le__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">:</span> <span class="nb">object</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="n">CompactGenome</span><span class="p">):</span>
            <span class="k">return</span> <span class="nb">sorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">items</span><span class="p">())</span> <span class="o">&lt;=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">other</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">items</span><span class="p">())</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">NotImplementedError</span>

    <span class="k">def</span> <span class="fm">__lt__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">:</span> <span class="nb">object</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="n">CompactGenome</span><span class="p">):</span>
            <span class="k">return</span> <span class="nb">sorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">items</span><span class="p">())</span> <span class="o">&lt;</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">other</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">items</span><span class="p">())</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">NotImplementedError</span>

    <span class="k">def</span> <span class="fm">__gt__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">:</span> <span class="nb">object</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
        <span class="k">return</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="fm">__le__</span><span class="p">(</span><span class="n">other</span><span class="p">)</span>

    <span class="k">def</span> <span class="fm">__ge__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">:</span> <span class="nb">object</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
        <span class="k">return</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="fm">__lt__</span><span class="p">(</span><span class="n">other</span><span class="p">)</span>

    <span class="k">def</span> <span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;CompactGenome(</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="si">}</span><span class="s2">,&quot;</span>
            <span class="sa">f</span><span class="s2">&quot; &lt;reference sequence str with id:</span><span class="si">{</span><span class="nb">id</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reference</span><span class="p">)</span><span class="si">}</span><span class="s2">&gt;)&quot;</span>
        <span class="p">)</span>

    <span class="k">def</span> <span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;CompactGenome[</span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutations_as_strings</span><span class="p">())</span><span class="si">}</span><span class="s2">]&quot;</span>

<div class="viewcode-block" id="CompactGenome.get_site">
<a class="viewcode-back" href="../../stubs/historydag.compact_genome.html#historydag.compact_genome.CompactGenome.get_site">[docs]</a>
    <span class="k">def</span> <span class="nf">get_site</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">site</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get the base at the provided (one-based) site index.&quot;&quot;&quot;</span>
        <span class="n">mut</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">site</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">mut</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference</span><span class="p">[</span><span class="n">site</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">mut</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span></div>


<div class="viewcode-block" id="CompactGenome.mutations_as_strings">
<a class="viewcode-back" href="../../stubs/historydag.compact_genome.html#historydag.compact_genome.CompactGenome.mutations_as_strings">[docs]</a>
    <span class="k">def</span> <span class="nf">mutations_as_strings</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return mutations as a tuple of strings of the format &#39;&lt;reference</span>
<span class="sd">        base&gt;&lt;index&gt;&lt;new base&gt;&#39;, sorted by index.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">tuple</span><span class="p">(</span>
            <span class="p">(</span><span class="n">startbase</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">index</span><span class="p">)</span> <span class="o">+</span> <span class="n">endbase</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="p">(</span><span class="n">startbase</span><span class="p">,</span> <span class="n">endbase</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">t</span><span class="p">:</span> <span class="n">t</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="p">)</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="CompactGenome.mutate">
<a class="viewcode-back" href="../../stubs/historydag.compact_genome.html#historydag.compact_genome.CompactGenome.mutate">[docs]</a>
    <span class="k">def</span> <span class="nf">mutate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">mutstring</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">reverse</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Apply a mutstring such as &#39;A110G&#39; to this compact genome.</span>

<span class="sd">        In this example, A is the old base, G is the new base, and 110 is the 1-based</span>
<span class="sd">        index of the mutation in the sequence. Returns the new</span>
<span class="sd">        CompactGenome, and prints a warning if the old base doesn&#39;t</span>
<span class="sd">        match the recorded old base in this compact genome.</span>

<span class="sd">        Args:</span>
<span class="sd">            mutstring: The mutation to apply</span>
<span class="sd">            reverse: Apply the mutation in reverse, such as when the provided mutation</span>
<span class="sd">                describes how to achieve this CompactGenome from the desired CompactGenome.</span>
<span class="sd">        Returns:</span>
<span class="sd">            The new CompactGenome</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">oldbase</span> <span class="o">=</span> <span class="n">mutstring</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">newbase</span> <span class="o">=</span> <span class="n">mutstring</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">reverse</span><span class="p">:</span>
            <span class="n">oldbase</span><span class="p">,</span> <span class="n">newbase</span> <span class="o">=</span> <span class="n">newbase</span><span class="p">,</span> <span class="n">oldbase</span>
        <span class="n">idx</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">mutstring</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
        <span class="n">ref_base</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference</span><span class="p">[</span><span class="n">idx</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span>
        <span class="n">idx_present</span> <span class="o">=</span> <span class="n">idx</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutations</span>
        <span class="k">if</span> <span class="n">idx_present</span><span class="p">:</span>
            <span class="n">old_recorded_base</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="p">[</span><span class="n">idx</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">old_recorded_base</span> <span class="o">=</span> <span class="n">ref_base</span>

        <span class="k">if</span> <span class="n">oldbase</span> <span class="o">!=</span> <span class="n">old_recorded_base</span><span class="p">:</span>
            <span class="n">warn</span><span class="p">(</span><span class="s2">&quot;recorded old base in sequence doesn&#39;t match old base&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">ref_base</span> <span class="o">==</span> <span class="n">newbase</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">idx_present</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">CompactGenome</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="n">idx</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">return</span> <span class="bp">self</span>
        <span class="k">return</span> <span class="n">CompactGenome</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">idx</span><span class="p">,</span> <span class="p">(</span><span class="n">ref_base</span><span class="p">,</span> <span class="n">newbase</span><span class="p">)),</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="CompactGenome.apply_muts_raw">
<a class="viewcode-back" href="../../stubs/historydag.compact_genome.html#historydag.compact_genome.CompactGenome.apply_muts_raw">[docs]</a>
    <span class="k">def</span> <span class="nf">apply_muts_raw</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">muts</span><span class="p">:</span> <span class="n">Sequence</span><span class="p">[</span><span class="nb">tuple</span><span class="p">]):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Apply the mutations from the sequence of tuples ``muts``.</span>

<span class="sd">        Each tuple should contain (one-based site, from_base, to_base)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Is this tested? What does raw mean? Are there checks? What&#39;s mutate</span>
        <span class="c1"># for?</span>
        <span class="n">res</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">site</span><span class="p">,</span> <span class="n">from_base</span><span class="p">,</span> <span class="n">to_base</span> <span class="ow">in</span> <span class="n">muts</span><span class="p">:</span>
            <span class="n">ref</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference</span><span class="p">[</span><span class="n">site</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">ref</span> <span class="o">!=</span> <span class="n">to_base</span><span class="p">:</span>
                <span class="n">res</span><span class="p">[</span><span class="n">site</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">ref</span><span class="p">,</span> <span class="n">to_base</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">res</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">site</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">CompactGenome</span><span class="p">(</span><span class="n">res</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference</span><span class="p">)</span></div>


<div class="viewcode-block" id="CompactGenome.apply_muts">
<a class="viewcode-back" href="../../stubs/historydag.compact_genome.html#historydag.compact_genome.CompactGenome.apply_muts">[docs]</a>
    <span class="k">def</span> <span class="nf">apply_muts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">muts</span><span class="p">:</span> <span class="n">Sequence</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">reverse</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">debug</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Apply a sequence of mutstrings like &#39;A110G&#39; to this compact genome.</span>

<span class="sd">        In this example, A is the old base, G is the new base, and 110 is the 1-based</span>
<span class="sd">        index of the mutation in the sequence. Returns the new</span>
<span class="sd">        CompactGenome, and prints a warning if the old base doesn&#39;t</span>
<span class="sd">        match the recorded old base in this compact genome.</span>

<span class="sd">        Args:</span>
<span class="sd">            muts: The mutations to apply, in the order they should be applied</span>
<span class="sd">            reverse: Apply the mutations in reverse, such as when the provided mutations</span>
<span class="sd">                describe how to achieve this CompactGenome from the desired CompactGenome.</span>
<span class="sd">                If True, the mutations in `muts` will also be applied in reversed order.</span>
<span class="sd">            debug: If True, each mutation is applied individually by</span>
<span class="sd">                :meth:`CompactGenome.apply_mut` and the from base is checked against the</span>
<span class="sd">                current recorded base at each site.</span>

<span class="sd">        Returns:</span>
<span class="sd">            The new CompactGenome</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">newcg</span> <span class="o">=</span> <span class="bp">self</span>
        <span class="k">if</span> <span class="n">reverse</span><span class="p">:</span>
            <span class="n">mod_func</span> <span class="o">=</span> <span class="nb">reversed</span>

            <span class="k">def</span> <span class="nf">rearrange_func</span><span class="p">(</span><span class="n">tup</span><span class="p">):</span>
                <span class="k">return</span> <span class="n">tup</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">tup</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">tup</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="k">def</span> <span class="nf">mod_func</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
                <span class="k">yield from</span> <span class="n">seq</span>

            <span class="k">def</span> <span class="nf">rearrange_func</span><span class="p">(</span><span class="n">tup</span><span class="p">):</span>
                <span class="k">return</span> <span class="n">tup</span>

        <span class="k">if</span> <span class="n">debug</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="n">mod_func</span><span class="p">(</span><span class="n">muts</span><span class="p">):</span>
                <span class="n">newcg</span> <span class="o">=</span> <span class="n">newcg</span><span class="o">.</span><span class="n">mutate</span><span class="p">(</span><span class="n">mut</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="n">reverse</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">newcg</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">apply_muts_raw</span><span class="p">(</span>
                <span class="n">rearrange_func</span><span class="p">(</span><span class="n">unpack_mut_string</span><span class="p">(</span><span class="n">mut</span><span class="p">))</span> <span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="n">mod_func</span><span class="p">(</span><span class="n">muts</span><span class="p">)</span>
            <span class="p">)</span>

        <span class="k">return</span> <span class="n">newcg</span></div>


<div class="viewcode-block" id="CompactGenome.to_sequence">
<a class="viewcode-back" href="../../stubs/historydag.compact_genome.html#historydag.compact_genome.CompactGenome.to_sequence">[docs]</a>
    <span class="k">def</span> <span class="nf">to_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Convert this CompactGenome to a full nucleotide sequence.&quot;&quot;&quot;</span>
        <span class="n">newseq</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">newseq</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reference</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="p">(</span><span class="n">ref_base</span><span class="p">,</span> <span class="n">newbase</span><span class="p">)</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">ref_base</span> <span class="o">!=</span> <span class="n">newseq</span><span class="p">[</span><span class="n">idx</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]:</span>
                <span class="n">warn</span><span class="p">(</span>
                    <span class="s2">&quot;CompactGenome.to_sequence warning: reference base doesn&#39;t match cg reference base&quot;</span>
                <span class="p">)</span>
            <span class="n">newseq</span><span class="p">[</span><span class="n">idx</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">newbase</span>
        <span class="k">return</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">newseq</span><span class="p">)</span></div>


<div class="viewcode-block" id="CompactGenome.mask_sites">
<a class="viewcode-back" href="../../stubs/historydag.compact_genome.html#historydag.compact_genome.CompactGenome.mask_sites">[docs]</a>
    <span class="k">def</span> <span class="nf">mask_sites</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sites</span><span class="p">,</span> <span class="n">one_based</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Remove any mutations on sites in `sites`, leaving the reference</span>
<span class="sd">        sequence unchanged.</span>

<span class="sd">        Args:</span>
<span class="sd">            sites: A collection of sites to be masked</span>
<span class="sd">            one_based: If True, the provided sites will be interpreted as one-based sites. Otherwise,</span>
<span class="sd">                they will be interpreted as 0-based sites.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">sites</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">sites</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">one_based</span><span class="p">:</span>

            <span class="k">def</span> <span class="nf">site_translate</span><span class="p">(</span><span class="n">site</span><span class="p">):</span>
                <span class="k">return</span> <span class="n">site</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="k">def</span> <span class="nf">site_translate</span><span class="p">(</span><span class="n">site</span><span class="p">):</span>
                <span class="k">return</span> <span class="n">site</span> <span class="o">-</span> <span class="mi">1</span>

        <span class="k">return</span> <span class="n">CompactGenome</span><span class="p">(</span>
            <span class="p">{</span>
                <span class="n">site</span><span class="p">:</span> <span class="n">data</span>
                <span class="k">for</span> <span class="n">site</span><span class="p">,</span> <span class="n">data</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
                <span class="k">if</span> <span class="n">site_translate</span><span class="p">(</span><span class="n">site</span><span class="p">)</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">sites</span>
            <span class="p">},</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">reference</span><span class="p">,</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="CompactGenome.superset_sites">
<a class="viewcode-back" href="../../stubs/historydag.compact_genome.html#historydag.compact_genome.CompactGenome.superset_sites">[docs]</a>
    <span class="k">def</span> <span class="nf">superset_sites</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sites</span><span class="p">,</span> <span class="n">new_reference</span><span class="p">,</span> <span class="n">one_based</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Do the opposite of `subset_sites`, adjusting site indices from</span>
<span class="sd">        indices in a sequence of variant sites, to indices in a sequence</span>
<span class="sd">        containing all sites.</span>

<span class="sd">        Args:</span>
<span class="sd">            sites: A sorted list of sites in the new_reference sequence which are represented</span>
<span class="sd">                by sites in the current compact genome&#39;s reference sequence</span>
<span class="sd">            new_reference: A new reference sequence</span>
<span class="sd">            one_based: Whether the sites in `sites` are one-based</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">one_based</span><span class="p">:</span>
            <span class="n">adjust</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">adjust</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="k">return</span> <span class="n">CompactGenome</span><span class="p">(</span>
            <span class="p">{</span><span class="n">sites</span><span class="p">[</span><span class="n">site</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">adjust</span><span class="p">:</span> <span class="n">mut</span> <span class="k">for</span> <span class="n">site</span><span class="p">,</span> <span class="n">mut</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">items</span><span class="p">()},</span>
            <span class="n">new_reference</span><span class="p">,</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="CompactGenome.subset_sites">
<a class="viewcode-back" href="../../stubs/historydag.compact_genome.html#historydag.compact_genome.CompactGenome.subset_sites">[docs]</a>
    <span class="k">def</span> <span class="nf">subset_sites</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sites</span><span class="p">,</span> <span class="n">one_based</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">new_reference</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Remove all but those sites in ``sites``, and adjust the reference</span>
<span class="sd">        sequence.</span>

<span class="sd">        Args:</span>
<span class="sd">            sites: A collection of sites to be kept</span>
<span class="sd">            one_based: If True, the provided sites will be interpreted as one-based sites. Otherwise,</span>
<span class="sd">                they will be interpreted as 0-based sites.</span>
<span class="sd">            new_reference: If provided, this new reference sequence will be used instead of</span>
<span class="sd">                computing the new reference sequence directly.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">one_based</span><span class="p">:</span>
            <span class="n">adjust</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">adjust</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="n">site_map</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">old_site</span> <span class="o">+</span> <span class="n">adjust</span><span class="p">:</span> <span class="n">new_site</span>
            <span class="k">for</span> <span class="n">new_site</span><span class="p">,</span> <span class="n">old_site</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">sites</span><span class="p">),</span> <span class="n">start</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="p">}</span>
        <span class="n">result</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">site_map</span><span class="p">[</span><span class="n">old_site</span><span class="p">]:</span> <span class="n">state</span>
            <span class="k">for</span> <span class="n">old_site</span><span class="p">,</span> <span class="n">state</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">old_site</span> <span class="o">-</span> <span class="n">adjust</span> <span class="ow">in</span> <span class="n">sites</span>
        <span class="p">}</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">result</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="p">):</span>
            <span class="n">warn</span><span class="p">(</span><span class="s2">&quot;Sites where cg differed from reference removed.&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">new_reference</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">new_reference</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">reference</span><span class="p">[</span><span class="n">site</span> <span class="o">+</span> <span class="n">adjust</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">sites</span><span class="p">)</span>
            <span class="p">)</span>

        <span class="k">return</span> <span class="n">CompactGenome</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="n">new_reference</span><span class="p">)</span></div>


<div class="viewcode-block" id="CompactGenome.remove_sites">
<a class="viewcode-back" href="../../stubs/historydag.compact_genome.html#historydag.compact_genome.CompactGenome.remove_sites">[docs]</a>
    <span class="k">def</span> <span class="nf">remove_sites</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sites</span><span class="p">,</span> <span class="n">one_based</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">new_reference</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Remove all sites in ``sites``, and adjust the reference sequence.</span>

<span class="sd">        Args:</span>
<span class="sd">            sites: A collection of sites to be removed</span>
<span class="sd">            one_based: If True, the provided sites will be interpreted as one-based sites. Otherwise,</span>
<span class="sd">                they will be interpreted as 0-based sites.</span>
<span class="sd">            new_reference: If provided, this new reference sequence will be used instead of</span>
<span class="sd">                computing the new reference sequence directly.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">one_based</span><span class="p">:</span>
            <span class="n">site_adjust</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">site_adjust</span> <span class="o">=</span> <span class="mi">1</span>

        <span class="k">if</span> <span class="n">new_reference</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">new_reference</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="n">base</span>
                <span class="k">for</span> <span class="n">site</span><span class="p">,</span> <span class="n">base</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reference</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
                <span class="k">if</span> <span class="p">(</span><span class="n">site</span> <span class="o">-</span> <span class="n">site_adjust</span><span class="p">)</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">sites</span>
            <span class="p">)</span>

        <span class="k">return</span> <span class="n">CompactGenome</span><span class="p">(</span>
            <span class="p">{</span>
                <span class="n">mod_site</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="p">[</span><span class="n">site</span><span class="p">]</span>
                <span class="k">for</span> <span class="n">mod_site</span><span class="p">,</span> <span class="n">site</span> <span class="ow">in</span> <span class="n">_iter_adjusted_sites</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="n">sites</span><span class="p">,</span> <span class="n">site_adjust</span>
                <span class="p">)</span>
            <span class="p">},</span>
            <span class="n">new_reference</span><span class="p">,</span>
        <span class="p">)</span></div>
</div>



<div class="viewcode-block" id="unpack_mut_string">
<a class="viewcode-back" href="../../stubs/historydag.compact_genome.html#historydag.compact_genome.unpack_mut_string">[docs]</a>
<span class="k">def</span> <span class="nf">unpack_mut_string</span><span class="p">(</span><span class="n">mut</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns (one-based site, from_base, to_base)&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">int</span><span class="p">(</span><span class="n">mut</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]),</span> <span class="n">mut</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">mut</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span></div>



<span class="k">def</span> <span class="nf">_iter_adjusted_sites</span><span class="p">(</span><span class="n">recorded_sites</span><span class="p">,</span> <span class="n">removed_sites</span><span class="p">,</span> <span class="n">site_adjust</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Adjusts recorded_sites if removed_sites are removed.</span>

<span class="sd">    For each one, returns a pair (modified site, unmodified site).</span>
<span class="sd">    site_adjust is the amount by which removed_sites base index is</span>
<span class="sd">    smaller than recorded_sites&#39; base index.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">all_sites</span> <span class="o">=</span> <span class="p">{</span><span class="n">site</span><span class="p">:</span> <span class="kc">False</span> <span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="n">recorded_sites</span><span class="p">}</span>
    <span class="n">all_sites</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="n">site</span> <span class="o">+</span> <span class="n">site_adjust</span><span class="p">:</span> <span class="kc">True</span> <span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="n">removed_sites</span><span class="p">})</span>
    <span class="n">shift</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">site</span><span class="p">,</span> <span class="n">removed</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">all_sites</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
        <span class="k">if</span> <span class="n">removed</span><span class="p">:</span>
            <span class="n">shift</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">yield</span> <span class="n">site</span> <span class="o">-</span> <span class="n">shift</span><span class="p">,</span> <span class="n">site</span>


<div class="viewcode-block" id="compact_genome_from_sequence">
<a class="viewcode-back" href="../../stubs/historydag.compact_genome.html#historydag.compact_genome.compact_genome_from_sequence">[docs]</a>
<span class="k">def</span> <span class="nf">compact_genome_from_sequence</span><span class="p">(</span><span class="n">sequence</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">reference</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create a CompactGenome from a sequence and a reference sequence.</span>

<span class="sd">    Args:</span>
<span class="sd">        sequence: the sequence to be represented by a CompactGenome</span>
<span class="sd">        reference: the reference sequence for the CompactGenome</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">cg</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">zero_idx</span> <span class="o">+</span> <span class="mi">1</span><span class="p">:</span> <span class="p">(</span><span class="n">old_base</span><span class="p">,</span> <span class="n">new_base</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">zero_idx</span><span class="p">,</span> <span class="p">(</span><span class="n">old_base</span><span class="p">,</span> <span class="n">new_base</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">reference</span><span class="p">,</span> <span class="n">sequence</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">old_base</span> <span class="o">!=</span> <span class="n">new_base</span>
    <span class="p">}</span>
    <span class="k">return</span> <span class="n">CompactGenome</span><span class="p">(</span><span class="n">cg</span><span class="p">,</span> <span class="n">reference</span><span class="p">)</span></div>



<div class="viewcode-block" id="cg_diff">
<a class="viewcode-back" href="../../stubs/historydag.compact_genome.html#historydag.compact_genome.cg_diff">[docs]</a>
<span class="k">def</span> <span class="nf">cg_diff</span><span class="p">(</span><span class="n">parent_cg</span><span class="p">:</span> <span class="n">CompactGenome</span><span class="p">,</span> <span class="n">child_cg</span><span class="p">:</span> <span class="n">CompactGenome</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Yields mutations in the format (parent_nuc, child_nuc, one-based</span>
<span class="sd">    sequence_index) distinguishing two compact genomes, such that applying the</span>
<span class="sd">    resulting mutations to `parent_cg` would yield `child_cg`&quot;&quot;&quot;</span>
    <span class="n">keys</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">parent_cg</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span> <span class="o">|</span> <span class="nb">set</span><span class="p">(</span><span class="n">child_cg</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
    <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">keys</span><span class="p">:</span>
        <span class="n">parent_base</span> <span class="o">=</span> <span class="n">parent_cg</span><span class="o">.</span><span class="n">get_site</span><span class="p">(</span><span class="n">key</span><span class="p">)</span>
        <span class="n">child_base</span> <span class="o">=</span> <span class="n">child_cg</span><span class="o">.</span><span class="n">get_site</span><span class="p">(</span><span class="n">key</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">parent_base</span> <span class="o">!=</span> <span class="n">child_base</span><span class="p">:</span>
            <span class="k">yield</span> <span class="p">(</span><span class="n">parent_base</span><span class="p">,</span> <span class="n">child_base</span><span class="p">,</span> <span class="n">key</span><span class="p">)</span></div>



<div class="viewcode-block" id="ambiguous_cg_diff">
<a class="viewcode-back" href="../../stubs/historydag.compact_genome.html#historydag.compact_genome.ambiguous_cg_diff">[docs]</a>
<span class="k">def</span> <span class="nf">ambiguous_cg_diff</span><span class="p">(</span>
    <span class="n">parent_cg</span><span class="p">:</span> <span class="n">CompactGenome</span><span class="p">,</span>
    <span class="n">child_cg</span><span class="p">:</span> <span class="n">CompactGenome</span><span class="p">,</span>
    <span class="n">transition_model</span><span class="o">=</span><span class="n">default_nt_transitions</span><span class="p">,</span>
    <span class="n">randomize</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Yields a minimal collection of mutations in the format (parent_nuc,</span>
<span class="sd">    child_nuc, one-based sequence_index) distinguishing two compact genomes,</span>
<span class="sd">    such that applying the resulting mutations to `parent_cg` would yield a</span>
<span class="sd">    compact genome compatible with the possibly ambiguous `child_cg`.</span>

<span class="sd">    If randomize is True, mutations will be randomized when there are</span>
<span class="sd">    multiple possible min-weight choices.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">parent_base</span><span class="p">,</span> <span class="n">child_base</span><span class="p">,</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">cg_diff</span><span class="p">(</span><span class="n">parent_cg</span><span class="p">,</span> <span class="n">child_cg</span><span class="p">):</span>
        <span class="n">nbase</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">transition_model</span><span class="o">.</span><span class="n">min_character_mutation</span><span class="p">(</span>
            <span class="n">parent_base</span><span class="p">,</span>
            <span class="n">child_base</span><span class="p">,</span>
            <span class="n">site</span><span class="o">=</span><span class="n">key</span><span class="p">,</span>
            <span class="n">randomize</span><span class="o">=</span><span class="n">randomize</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">parent_base</span> <span class="o">!=</span> <span class="n">nbase</span><span class="p">:</span>
            <span class="k">yield</span> <span class="p">(</span><span class="n">parent_base</span><span class="p">,</span> <span class="n">nbase</span><span class="p">,</span> <span class="n">key</span><span class="p">)</span></div>



<div class="viewcode-block" id="reconcile_cgs">
<a class="viewcode-back" href="../../stubs/historydag.compact_genome.html#historydag.compact_genome.reconcile_cgs">[docs]</a>
<span class="k">def</span> <span class="nf">reconcile_cgs</span><span class="p">(</span>
    <span class="n">cg_list</span><span class="p">,</span> <span class="n">check_references</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ambiguitymap</span><span class="o">=</span><span class="n">standard_nt_ambiguity_map</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a compact genome containing ambiguous bases, representing the</span>
<span class="sd">    least ambiguous sequence of which all provided cgs in `cg_list` are</span>
<span class="sd">    resolutions. Also returns a flag indicating whether the resulting CG</span>
<span class="sd">    contains ambiguities.</span>

<span class="sd">    If `check_references` is False, reference sequences will be assumed equal.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">ambiguous_flag</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cg_list</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">(</span><span class="n">cg_list</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">ambiguous_flag</span><span class="p">)</span>

    <span class="n">cg_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">cg_list</span><span class="p">)</span>
    <span class="n">model_cg</span> <span class="o">=</span> <span class="n">cg_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">reference</span> <span class="o">=</span> <span class="n">model_cg</span><span class="o">.</span><span class="n">reference</span>
    <span class="k">if</span> <span class="n">check_references</span><span class="p">:</span>
        <span class="k">assert</span> <span class="nb">all</span><span class="p">(</span><span class="n">cg</span><span class="o">.</span><span class="n">reference</span> <span class="o">==</span> <span class="n">reference</span> <span class="k">for</span> <span class="n">cg</span> <span class="ow">in</span> <span class="n">cg_list</span><span class="p">)</span>

    <span class="c1"># This can almost certainly be made more efficient</span>
    <span class="n">difference_sites</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">ocg</span> <span class="ow">in</span> <span class="n">cg_list</span><span class="p">[</span><span class="mi">1</span><span class="p">:]:</span>
        <span class="n">diff</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">cg_diff</span><span class="p">(</span><span class="n">model_cg</span><span class="p">,</span> <span class="n">ocg</span><span class="p">))</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">diff</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">ambiguous_flag</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">for</span> <span class="n">parent_nuc</span><span class="p">,</span> <span class="n">child_nuc</span><span class="p">,</span> <span class="n">one_idx</span> <span class="ow">in</span> <span class="n">diff</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">one_idx</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">difference_sites</span><span class="p">:</span>
                <span class="n">difference_sites</span><span class="p">[</span><span class="n">one_idx</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="n">parent_nuc</span><span class="p">,</span> <span class="n">child_nuc</span><span class="p">}</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># Parent nuc must already be added!</span>
                <span class="n">difference_sites</span><span class="p">[</span><span class="n">one_idx</span><span class="p">]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">child_nuc</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">process_charset</span><span class="p">(</span><span class="n">charset</span><span class="p">):</span>
        <span class="n">diff</span> <span class="o">=</span> <span class="n">charset</span> <span class="o">-</span> <span class="n">ambiguitymap</span><span class="o">.</span><span class="n">bases</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">diff</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">charset</span><span class="o">.</span><span class="n">intersection_update</span><span class="p">(</span><span class="n">ambiguitymap</span><span class="o">.</span><span class="n">bases</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">char</span> <span class="ow">in</span> <span class="n">diff</span><span class="p">:</span>
                <span class="n">charset</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">ambiguitymap</span><span class="p">[</span><span class="n">char</span><span class="p">])</span>
        <span class="k">return</span> <span class="nb">frozenset</span><span class="p">(</span><span class="n">charset</span><span class="p">)</span>

    <span class="n">mutstring_list</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">model_cg</span><span class="o">.</span><span class="n">get_site</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
        <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
        <span class="o">+</span> <span class="n">ambiguitymap</span><span class="o">.</span><span class="n">reversed</span><span class="p">[</span><span class="n">process_charset</span><span class="p">(</span><span class="n">charset</span><span class="p">)]</span>
        <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">charset</span> <span class="ow">in</span> <span class="n">difference_sites</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
    <span class="p">]</span>

    <span class="k">return</span> <span class="p">(</span><span class="n">model_cg</span><span class="o">.</span><span class="n">apply_muts</span><span class="p">(</span><span class="n">mutstring_list</span><span class="p">),</span> <span class="n">ambiguous_flag</span><span class="p">)</span></div>



<div class="viewcode-block" id="read_alignment">
<a class="viewcode-back" href="../../stubs/historydag.compact_genome.html#historydag.compact_genome.read_alignment">[docs]</a>
<span class="k">def</span> <span class="nf">read_alignment</span><span class="p">(</span><span class="n">alignment_file</span><span class="p">,</span> <span class="n">reference_sequence</span><span class="p">:</span> <span class="n">CharacterSequence</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Read a fasta or vcf alignment and return a dictionary mapping sequence</span>
<span class="sd">    ID strings to CompactGenomes.</span>

<span class="sd">    Args:</span>
<span class="sd">        alignment_file: A file containing a fasta or vcf alignment. File format</span>
<span class="sd">            is determined by extension. `.fa`, `.fasta`, or `.vcf` are expected.</span>
<span class="sd">        reference_sequence: If a fasta file is provided, the first sequence in that</span>
<span class="sd">            file will be used as the compact genome reference sequence, unless one</span>
<span class="sd">            is explicitly provided to this keyword argument.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">extension</span> <span class="o">=</span> <span class="n">alignment_file</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">extension</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;fa&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">):</span>
        <span class="n">fasta_gen</span> <span class="o">=</span> <span class="n">read_fasta</span><span class="p">(</span><span class="n">alignment_file</span><span class="p">)</span>
        <span class="n">cg_dict</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">if</span> <span class="n">reference_sequence</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># Now we just have to add an empty CG to the alignment</span>
            <span class="n">refseq_id</span><span class="p">,</span> <span class="n">reference_sequence</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">fasta_gen</span><span class="p">)</span>
            <span class="n">cg_dict</span><span class="p">[</span><span class="n">refseq_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">CompactGenome</span><span class="p">({},</span> <span class="n">reference_sequence</span><span class="p">)</span>
        <span class="n">cg_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
            <span class="p">(</span><span class="n">seqid</span><span class="p">,</span> <span class="n">compact_genome_from_sequence</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">reference_sequence</span><span class="p">))</span>
            <span class="k">for</span> <span class="n">seqid</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">fasta_gen</span>
        <span class="p">)</span>
    <span class="k">elif</span> <span class="n">extension</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;vcf&quot;</span><span class="p">,):</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Unrecognized extension &#39;.</span><span class="si">{</span><span class="n">extension</span><span class="si">}</span><span class="s2">&#39;. Provide a .fa, .fasta, or .vcf file.&quot;</span>
        <span class="p">)</span>
    <span class="k">return</span> <span class="n">cg_dict</span></div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2020, Will Dumm.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>